syncer/sharding-meta/shardmeta.go

```diff
@@ -0,0 +1,232 @@
+	if exists {
+		return index == meta.activeIdx, nil
+	}
+
+	if source, ok := meta.sources[item.Source]; !ok {
+		meta.sources[item.Source] = &ShardingSequence{Items: []*DDLItem{item}}
+	} else {
+		source.Items = append(source.Items, item)
+	}
+
+	found := false
+	for _, globalItem := range meta.global.Items {
+		if utils.CompareShardingDDLs(item.DDLs, globalItem.DDLs) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		meta.global.Items = append(meta.global.Items, item)
+	}
+
+	global, source := meta.global, meta.sources[item.Source]
+	if !source.IsPrefixSequence(global) {
+		return false, errors.Errorf("wrong sql sequence of source %+v, global is %+v", source.Items, global.Items)
+	}
+
+	return index == meta.activeIdx, nil
+}
+
+// GetActiveDDLItem returns the source table's active DDLItem
+// if in DDL unsynced procedure, the active DDLItem means the syncing DDL
+// if in re-sync procedure, the active DDLItem means the next syncing DDL in DDL syncing sequence, may be nil
+func (meta *ShardingMeta) GetActiveDDLItem(tableSource string) *DDLItem {
+	source, ok := meta.sources[tableSource]
+	if !ok {
+		return nil
+	}
+	if meta.activeIdx < len(source.Items) {
+		return source.Items[meta.activeIdx]
+	}
+	return nil
+}
+
+// InSequenceSharding returns whether in sequence sharding
+func (meta *ShardingMeta) InSequenceSharding() bool {
+	return meta.activeIdx < len(meta.global.Items)
+}
+
+// ResolveShardingDDL resolves one sharding DDL and increase activeIdx
+// if activeIdx equals to the length of global sharding sequence, it means all
+// sharding DDL in this ShardingMeta sequence is resolved and will reinitialize
+// the ShardingMeta, return true if all DDLs are resolved.
+func (meta *ShardingMeta) ResolveShardingDDL() bool {
+	meta.activeIdx++
+	if meta.activeIdx == len(meta.global.Items) {
+		meta.reinitialize()
+		return true
+	}
+	return false
+}
+
+// ActiveDDLFirstPos returns the first binlog position of active DDL
+func (meta *ShardingMeta) ActiveDDLFirstPos() (*mysql.Position, error) {
+	if meta.activeIdx >= len(meta.global.Items) {
+		return nil, errors.Errorf("activeIdx %d larger than global DDLItems: %v", meta.activeIdx, meta.global.Items)
+	}
+	return &meta.global.Items[meta.activeIdx].FirstPos, nil
+}
+
+// FlushData returns sharding meta flush SQL and args
+func (meta *ShardingMeta) FlushData(schema, table, sourceID, tableID string) ([]string, [][]interface{}) {
+	if len(meta.global.Items) == 0 {
+		sql2 := fmt.Sprintf("DELETE FROM `%s`.`%s` where source_id=? and table_id=?", schema, table)
+		args2 := []interface{}{sourceID, tableID}
+		return []string{sql2}, [][]interface{}{args2}
+	}
+	var (
+		sqls    = make([]string, 1+len(meta.sources))
+		args    = make([][]interface{}, 0, 1+len(meta.sources))
+		baseSQL = fmt.Sprintf("INSERT INTO `%s`.`%s` (`source_id`, `table_id`, `source`, `active`, `is_global`, `data`) VALUES(?,?,?,?,?,?) ON DUPLICATE KEY UPDATE `data`=?, `active`=?", schema, table)
```

Comment:
the `CREATE SCHEMA/TABLE` DDL statements in _sharding_group.go_, and these DML statements in this _shardmeta.go_, is there a method to put all of them together?

Reply:
as a matter of fact, `ShardingMeta` is a part of `ShardingGroup`, and `ShardingGroupKeeper` managers all sharding groups.
It is better to put them together as a whole package later.

Reply:
fine.